File: code/problems/longest_common_subsequence/s2.rb
def longest_common_subsequence(s1, s2)
  m, n = s1.length, s2.length
  lcs_lengths = Array.new(m + 1) { Array.new(n + 1, 0) }
  (1 .. m).each do |i|
    (1 .. n).each do |j|
      if s1[i-1] == s2[j-1]
        lcs_lengths[i][j] = 1 + lcs_lengths[i-1][j-1]
      else
        lcs_lengths[i][j] = [
          lcs_lengths[i-1][j],
          lcs_lengths[i][j-1]
        ].max
      end
    end
  end
  lcs_lengths[m][n]
end

File: code/problems/longest_common_subsequence/s2.js
export function longestCommonSubsequence(s1, s2) {
  const [m, n] = [s1.length, s2.length]
  const lcsLengths = Array(m + 1).fill().map(() => Array(n + 1).fill(0))
  for (let i = 1; i <= m; i++) {
    for(let j = 1; j <= n; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        lcsLengths[i][j] = 1 + lcsLengths[i - 1][j - 1]
      } else {
        lcsLengths[i][j] = Math.max(
          lcsLengths[i - 1][j],
          lcsLengths[i][j - 1]
        )
      }
    }
  }
  return lcsLengths[m][n]
}
